src/main/cpp/src/timezones.cu

```diff
@@ -0,0 +1,206 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <thrust/binary_search.h>
+
+#include "timezones.hpp"
+
+using column = cudf::column;
+using column_device_view = cudf::column_device_view;
+using lists_column_device_view = cudf::detail::lists_column_device_view;
+using size_type = cudf::size_type;
+using struct_view = cudf::struct_view;
+
+namespace {
+
+/**
+ * @brief Convert the timestamp value to either UTC or a specified timezone
+ *
+ * This device function uses a binary search to find the instant of the transition
+ * to find the right offset to do the transition.
+ *
+ * To transition to UTC: do a binary search on the tzInstant child column and subtract
+ * the offset
+ *
+ * To transition from UTC: do a binary search on the utcInstant child column and add
+ * the offset
+ *
+ * @tparam typestamp_type type of the input and output timestamp
+ * @param timestamp input timestamp
+ * @param transitions the list column of transitions to figure out the correct offset
+ *                    to adjust the timestamp. The type of the values in this column is
+ *                    LIST<STRUCT<utcInstant: int64, tzInstant: int64, utcOffset: int32>>
+ * @param tz_index the index of the specified zone id in the transitions table
+ * @param to_utc whether we are converting to UTC or converting to the timezone
+ */
+template <typename timestamp_type>
+__device__ timestamp_type convert_timestamp_timezone(timestamp_type const &timestamp,
+                                                     lists_column_device_view const &transitions,
+                                                     size_type tz_index, bool to_utc) {
+
+  using duration_type = typename timestamp_type::duration;
+  using cuda::std::chrono::duration_cast;
+
+  auto epoch_seconds =
+      static_cast<int64_t>(duration_cast<cudf::duration_s>(timestamp.time_since_epoch()).count());
+
+  auto const tz_transitions = cudf::list_device_view{transitions, tz_index};
+
+  auto size = tz_transitions.size();
+
+  cudf::device_span<int64_t const> transition_times(
+      &(transitions.child()
+            .child(to_utc ? 1 : 0)
+            .data<int64_t>()[tz_transitions.element_offset(0)]),
+      static_cast<size_t>(size));
+
+  auto idx = thrust::upper_bound(thrust::seq, transition_times.begin(), transition_times.end(),
+                                 epoch_seconds) -
+             transition_times.begin();
+
+  auto const list_offset = tz_transitions.element_offset(size_type(idx - 1));
```

Comment:
If we have to subtract one from `idx`, how about using `thrust::lower_bound` in the code above.

Reply:
It doesn't actually work for this algorithm. When the timestamp actually matches the instant that it is being compared to, we need that element in this binary search. `upper::bound` gives us the insertion point after this element, so we subtract `1`. When the timestamp doesn't match, then lower bound will still give us an insertion point after the element that we need to use. So we use upper bound, and always subtract `1` to get the element we want every time.

Reply:
Also, to be clear the nature of the data in this list is that it will never have repeating elements (since it's a list of timestamps for each transition, and there is only 1 transition at any given time).

Reply:
>  whereas lower::bound gives us the insertion point before this element

No not that. Indeed, `lower_bound` gives the position "no less than" the search element. In our use case, we don't have duplicate elements thus it will give the exact position of the element. And that is also the same as `upper_bound - 1`. So I think we can actually use `lower_bound` here without `-1`.

Reply:
Found this: 
![image](https://github.com/NVIDIA/spark-rapids-jni/assets/7416935/dd349e02-d42e-4bd1-95c4-3eec015ec599)